components/layout_2020/table/mod.rs

```diff
@@ -2,12 +2,66 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
 
-//! HTML Tables (╯°□°)╯︵ ┻━┻.
+//! # HTML Tables (╯°□°)╯︵ ┻━┻
 //!
-//! See <https://html.spec.whatwg.org/multipage/table-processing-model> and
-//! <https://drafts.csswg.org/css-tables>.  This is heavily based on the latter specification, but
-//! note that it is still an Editor's Draft, so there is no guarantee that what is implemented here
-//! matches other browsers or the current specification.
+//! This implementation is based on the [table section of the HTML 5
+//! Specification](https://html.spec.whatwg.org/multipage/#tables), the draft [CSS Table Module Level
+//! 3](https://drafts.csswg.org/css-tables) and the [LayoutNG implementation of
+//! tables](https://source.chromium.org/chromium/chromium/src/third_party/+/main:blink/renderer/core/layout/table)
+//! in Blink. In general, the draft specification differs greatly from what other browsers do, so we
+//! generally follow LayoutNG when in question.
+//!
+//! Table layout is divided into two phases:
+//!
+//! 1. Box Tree Construction
+//! 2. Fragment Tree Construction
+//!
+//! ## Box Tree Construction
+//!
+//! During box tree construction, table layout (`construct.rs`) will traverse the DOM and construct
+//! the basic box tree representation of a table, using the structs defined in this file ([`Table`],
+//! [`TableTrackGroup`], [`TableTrack`], etc). When processing the DOM, elements are handled
+//! differently depending on their `display` value. For instance, an element with `display:
+//! table-cell` is treated as a table cell. These display values are automatically set for HTML
+//! elements by the user agent stylesheet.
+//!
+//! Inside each table is an array of [`TableSlot`]. A table slot represents either a cell, a cell
+//! location occupied by a cell originating from another place in the table, or is empty. In
+//! addition, when there are table model errors, a slot may spanned by more than one cell. In those
+//! cases, the top-left-most cell wins.
```

Comment:
It's not much clear what "top-left-most cell wins" means.

Reply:
I've changed this to: 

```rust
//! addition, when there are table model errors, a slot may spanned by more than one cell. In those
//! cases, the cell that comes first in DOM order will cover the slot.
```

Reply:
But isn't the slot covered by all the cells? Like

```html
<table border="1">
<tr>
  <td>aaaa</td>
  <td rowspan="2" style="background: rgba(255, 0, 255, 0.5)">bbb</td>
</tr><tr>
  <td colspan="2" style="background: rgba(0, 255, 255, 0.5)">ccc ccc ccc</td>
</tr>
</table>
```

![](https://github.com/servo/servo/assets/7477678/e6440490-9c96-4870-860b-2f8b67381353)

Reply:
Huh. Interesting. From https://html.spec.whatwg.org/multipage/tables.html#table-model-error:

"If any of the slots involved already had a [cell](https://html.spec.whatwg.org/multipage/tables.html#concept-cell) covering them, then this is a [table model error](https://html.spec.whatwg.org/multipage/tables.html#table-model-error). Those slots now have two cells overlapping."

I'm not sure how to interpret this. It looks like our layout differs from Firefox in that case.

Reply:
I've just removed this sentence for now.

Reply:
The difference from Firefox is that our intrinsic sizing is broken for `colspan="2"` so the columns are too narrow, but the overlapping seems the same.

Reply:
Sorry, I also noticed that the baseline was different, but you're right that it isn't related to the overlapping.